optuna/terminator/improvement/evaluator.py

```diff
@@ -128,3 +128,53 @@ def _validate_input(
                 "The intersection search space is empty. This condition is not supported by "
                 f"{cls.__name__}."
             )
+
+
+@experimental_class("3.2.0")
+class BestValueStagnationEvaluator(BaseImprovementEvaluator):
+    """
+    Evaluates the stagnation period of the best value in an optimization process.
+
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed toevaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
+
+    Args:
+        - max_stagnation_trials (int): The maximum number of trials allowed for stagnation.
+
+    Example Usage:
+        >>> evaluator = BestValueStagnationEvaluator(max_stagnation_trials=30)
+        >>> stagnated_trials = evaluator.evaluate()
+    """
+
+    def __init__(
+        self,
+        max_stagnation_trials: int = 30,
+    ):
+        if max_stagnation_trials < 0:
+            raise ValueError("The number of tolerance of step must not be negative.")
+        self._max_stagnation_trials = max_stagnation_trials
+
+    def evaluate(
+        self,
+        trials: List[FrozenTrial],
+        study_direction: StudyDirection,
+    ) -> int:
```

Comment:
A nit, but we want to follow the return type of the base class.
```suggestion
    ) -> float:
```